src/ir/properties.h

```diff
@@ -362,6 +362,83 @@ inline Expression* getFallthrough(
   }
 }
 
+// Look at all the intermediate fallthrough expressions and return the most
+// precise type we know this value will have.
```

Comment:
It would be good to comment how this handles the case of not having a most precise type. From the code it looks like nullability and heap types are treated independently, but the problem can happen among heap types too? And the code looks like it does a greedy search, so it might not find the "deepest" result,
```
    A
   / \
  B   C
 /
D
```
Here we might return C when D might be better for the caller. Though even between B and C it isn't clear which the caller would prefer, and the old code considered them both - can't that be better?

Reply:
If we see a value typed with both `B` (or `D`) and `C`, we know the value cannot possibly exist, so if we observe that, we can jump straight to the bottom type. The commit I just pushed now optimizes this case.